docs/proposals/cluster-autoscaling.md

```diff
@@ -0,0 +1,314 @@
+                  |                                  |
+                  | +--------+ +--------+ +--------+ |
+                  | |        | |        | |        | |
+                  | | Node1  | | Node2  | | NodeX  | |
+                  | |        | |        | |        | +<--+
+                  | +--------+ +--------+ +--------+ |   |
+                  | |cAdvisor| |cAdvisor| |cAdvisor| |   |
+                  | +--------+ +--------+ +----+---+ |   |
+                  |     |          |           |     |   |
+                  +-----|----------|-----------|-----+   |
+                        |          |           |         |
+                        |      +---v----+      |         |
+                        +----> |Heapster| <----+         |
+                               +---+----+                |
+                                   |                     |
+   Kubernetes                      |                     |
+-----------------------------------|---------------------|------
+   Google                          |                     |
+                                   |                     |
+                            +------v-----+       +-------+------+
+                            |Google Cloud|       |MIG Definition|
+                            | Monitoring |       +-------^------+
+                            +------------+               |
+                                   |                     |
+                                   |                     |
+                            +------v-----+               |
+                +------+    |Google Cloud+---------------+
+                |Config|--->| Autoscaler |
+                +------+    +------------+
+```
+
+CAdvisor instances gather metrics for nodes they are running on. Periodically Heapster (a service running
+on one of the nodes) queries all of the available nodes/cAdvisors, aggregates the metrics and pushes the
+usage to Google Cloud Monitoring (Heapster also takes Pod requests from apiserver but this fact is not
+depicted on the diagram). Google Cloud Autoscaler checks the metrics in GCM, compares the values
+and min/max cluster sizes from the config (that is passed through google-specific gcloud tool or web ui,
+and currently set up in kube-up/push) and updates the number of number of nodes in MIG (Managed Instance Group)
+definition. MIG then starts (or stops) the nodes.
+
+
+The current solution has a bunch of issues:
+
+* (I1) [GCE specific API] Autoscaler configuration is Google-specific thus on other cloud providers the configuration may look completely different.
+* (I2) [GCE specific logic] Autoscaler logic (scale on a custom metric) is Google-specific - it might be hard to get it working on AWS or elsewhere
+because their autoscaler either does not allow to use custom metric or interprets/acts differently.
+* (I3) [Works only for well formed pods] Autoscaler guarantee scheduling only the pods that are smaller than
```

Comment:
This needs more overview before you dive into the details. How about starting with something like: Autoscaler grows the cluster only in reaction to resource consumption of pods that are already running, i.e. it does not take into account pods that are pending. It still tends to increase the probability that a pod will schedule, by keeping the cluster "not too full" (reasonable amount of free resources), but this is not guaranteed -- for example, the resources might be too fragmented to meet the pod's requirements, or the pod might have a node selector that can't be satisfied by any of the nodes that have sufficient free resources. More specifically on the first point, cluster autoscaler guarantees scheduling only the pods that are smaller than ... [rest of what you had]"